src/ast/passes/semantic_analyser.cpp

```diff
@@ -216,6 +216,303 @@ class SemanticAnalyser : public Visitor<SemanticAnalyser> {
 
 } // namespace
 
+struct assign_checks {
+  bool want_map = false;
+  bool want_var = false;
+  bool want_map_key = false;
+};
+
+// clang-format off
+static const std::map<std::string, assign_checks> CALL_ASSIGN_SPEC = {
+  { "avg", { .want_map = true } },
+  { "cat", {} },
+  { "clear", {} },
+  { "count", { .want_map = true } },
+  { "debugf", {} },
+  { "delete", {} },
+  { "exit", {} },
+  { "hist", { .want_map = true } },
+  { "join", {} },
+  { "lhist", { .want_map = true } },
+  { "max", { .want_map = true } },
+  { "min", { .want_map = true } },
+  { "override", {} },
+  { "print", {} },
+  { "printf", {} },
+  { "signal", {} },
+  { "skboutput", {.want_var = true} },
+  { "stats", { .want_map = true } },
+  { "sum", { .want_map = true } },
+  { "system", {} },
+  { "time", {} },
+  { "zero", {} },
+};
+
+struct arg_type_spec {
+  Type type = Type::integer;
+  bool literal = false;
+  // This means that this check is only valid if there are
+  // this exact number of args in the call
+  size_t nargs = 0;
```

Comment:
This is a bit confusing (although the comment helps). I'd either consider renaming the field or dropping it completely as it's only applicable in 3 cases. And we still have other places where we need to call `check_arg` explicitly (like kstack/ustack) so maybe let's just handle these cases where the check depends on the call context separately.